File: modules/uc-engine/ios/EngineDiagnosticBridge.swift
import Foundation

// Fixed binding fields only; never serializes arbitrary errors or attributes.
public enum EngineDiagnosticBridge {
  private static var source: BindingHostDiagnosticSource {
    Bundle.main.bundleIdentifier?.hasSuffix(".Keyboard") == true ? .keyboardExtension : .application
  }
  public static func register() {
    try? registerHostDiagnosticSource(source: source, capability: .partial)
  }
  public static func record(_ event: BindingHostDiagnosticEvent) {
    _ = try? recordHostDiagnostic(source: source, event: event)
  }
  public static func begin(_ action: BindingHostDiagnosticAction) -> String? {
    (try? recordHostDiagnostic(source: source, event: .begin(action: action)))?.token
  }
  public static func finish(_ token: String?, succeeded: Bool) {
    guard let token else { return }
    record(.finish(token: token, outcome: succeeded ? .completed : .failed(reason: .unknown)))
  }
  public static func observe<Result>(_ action: BindingHostDiagnosticAction, _ operation: () throws -> Result) rethrows -> Result {
    let token = begin(action)
    do { let result = try operation(); finish(token, succeeded: true); return result }
    catch { finish(token, succeeded: false); throw error }
  }

  public static func map(_ value: BindingLocalCaptureStatus) -> [String: Any] {
    [
      "mode": String(describing: value.mode),
      "captureId": value.captureId.map { $0 as Any } ?? NSNull(),
      "remainingMs": NSNumber(value: value.remainingMs),
      "startedAtUtc": value.startedAtUtc.map { $0 as Any } ?? NSNull(),
      "endReason": value.endReason.map { String(describing: $0) as Any } ?? NSNull(),
      "lastCaptureId": value.lastCaptureId.map { $0 as Any } ?? NSNull(),
      "revision": NSNumber(value: value.revision),
    ]
  }
  public static func map(_ value: BindingSourceCoverage) -> [String: Any] {
    [
      "source": String(describing: value.source),
      "capability": String(describing: value.capability),
      "collection": String(describing: value.collection),
      "observedCount": NSNumber(value: value.observedCount),
      "policyFilteredCount": NSNumber(value: value.policyFilteredCount),
    ]
  }
  public static func map(_ value: BindingFileSourceCounts) -> [String: Any] {
    [
      "source": String(describing: value.source),
      "acceptedCount": NSNumber(value: value.acceptedCount),
      "writtenCount": NSNumber(value: value.writtenCount),
      "queueDroppedCount": NSNumber(value: value.queueDroppedCount),
      "quotaDroppedCount": NSNumber(value: value.quotaDroppedCount),
      "writeFailedCount": NSNumber(value: value.writeFailedCount),
      "lastWrittenAtMs": value.lastWrittenAtMs.map { NSNumber(value: $0) } ?? NSNull(),
    ]
  }
  public static func map(_ value: BindingLocalDiagnosticStatus) -> [String: Any] {
    [
      "runId": value.runId,
      "capture": map(value.capture),
      "observedRecords": NSNumber(value: value.observedRecords),
      "policyFilteredRecords": NSNumber(value: value.policyFilteredRecords),
      "schemaRejectedRecords": NSNumber(value: value.schemaRejectedRecords),
      "correlationLimitedRecords": NSNumber(value: value.correlationLimitedRecords),
      "engineVersion": value.engineVersion,
      "sourceCommit": value.sourceCommit,
      "counterScope": value.counterScope,
      "sources": value.sources.map { map($0) },
      "localFile": String(describing: value.localFile),
      "closed": value.closed,
    ]
  }
  public static func map(_ value: BindingLocalDiagnosticExportReport) -> [String: Any] {
    [
      "flush": String(describing: value.flush),
      "status": map(value.status),
      "requestedAtUtc": value.requestedAtUtc,
      "completedAtUtc": value.completedAtUtc,
      "otherProcessesFlushed": value.otherProcessesFlushed,
      "files": value.files.map { map($0) },
    ]
  }
}

File: modules/uc-engine/android/src/main/java/expo/modules/ucengine/EngineDiagnosticBridge.kt
package expo.modules.ucengine

import uniffi.uc_engine_uniffi.*
import java.util.Locale

internal object EngineDiagnosticBridge {
  fun register() { runCatching { registerHostDiagnosticSource(BindingHostDiagnosticSource.APPLICATION, BindingSourceCapability.PARTIAL) } }
  fun record(event: BindingHostDiagnosticEvent) { runCatching { recordHostDiagnostic(BindingHostDiagnosticSource.APPLICATION, event) } }
  fun begin(action: BindingHostDiagnosticAction): String? = runCatching {
    recordHostDiagnostic(BindingHostDiagnosticSource.APPLICATION, BindingHostDiagnosticEvent.Begin(action)).token
  }.getOrNull()
  fun finish(token: String?, succeeded: Boolean) {
    if (token == null) return
    record(BindingHostDiagnosticEvent.Finish(token, if (succeeded) BindingHostDiagnosticOutcome.Completed else BindingHostDiagnosticOutcome.Failed(BindingHostDiagnosticFailure.UNKNOWN)))
  }
  fun <T> observe(action: BindingHostDiagnosticAction, operation: () -> T): T {
    val token = begin(action)
    return try { operation().also { finish(token, true) } } catch (error: Throwable) { finish(token, false); throw error }
  }

  private fun name(value: Enum<*>): String = value.name.lowercase(Locale.ROOT).split("_").let { parts -> parts.first() + parts.drop(1).joinToString("") { it.replaceFirstChar(Char::uppercaseChar) } }
  fun map(value: BindingStopCaptureResult): String = name(value)
  fun map(value: BindingLocalCaptureStatus): Map<String, Any?> = mapOf(
    "mode" to name(value.mode),
    "captureId" to value.captureId,
    "remainingMs" to value.remainingMs.toDouble(),
    "startedAtUtc" to value.startedAtUtc,
    "endReason" to value.endReason?.let { name(it) },
    "lastCaptureId" to value.lastCaptureId,
    "revision" to value.revision.toDouble(),
  )
  fun map(value: BindingSourceCoverage): Map<String, Any?> = mapOf(
    "source" to name(value.source),
    "capability" to name(value.capability),
    "collection" to name(value.collection),
    "observedCount" to value.observedCount.toDouble(),
    "policyFilteredCount" to value.policyFilteredCount.toDouble(),
  )
  fun map(value: BindingFileSourceCounts): Map<String, Any?> = mapOf(
    "source" to name(value.source),
    "acceptedCount" to value.acceptedCount.toDouble(),
    "writtenCount" to value.writtenCount.toDouble(),
    "queueDroppedCount" to value.queueDroppedCount.toDouble(),
    "quotaDroppedCount" to value.quotaDroppedCount.toDouble(),
    "writeFailedCount" to value.writeFailedCount.toDouble(),
    "lastWrittenAtMs" to value.lastWrittenAtMs?.toDouble(),
  )
  fun map(value: BindingLocalDiagnosticStatus): Map<String, Any?> = mapOf(
    "runId" to value.runId,
    "capture" to map(value.capture),
    "observedRecords" to value.observedRecords.toDouble(),
    "policyFilteredRecords" to value.policyFilteredRecords.toDouble(),
    "schemaRejectedRecords" to value.schemaRejectedRecords.toDouble(),
    "correlationLimitedRecords" to value.correlationLimitedRecords.toDouble(),
    "engineVersion" to value.engineVersion,
    "sourceCommit" to value.sourceCommit,
    "counterScope" to value.counterScope,
    "sources" to value.sources.map { map(it) },
    "localFile" to name(value.localFile),
    "closed" to value.closed,
  )
  fun map(value: BindingLocalDiagnosticExportReport): Map<String, Any?> = mapOf(
    "flush" to name(value.flush),
    "status" to map(value.status),
    "requestedAtUtc" to value.requestedAtUtc,
    "completedAtUtc" to value.completedAtUtc,
    "otherProcessesFlushed" to value.otherProcessesFlushed,
    "files" to value.files.map { map(it) },
  )
}
